tidy3d/components/medium.py

```diff
@@ -140,15 +301,114 @@ class NonlinearSusceptibility(NonlinearSpec):
+            self._validate_medium_freqs(medium, [])
 
-NonlinearSpecType = Union[NonlinearSusceptibility]
+
+NonlinearModelType = Union[NonlinearSusceptibility, TwoPhotonAbsorption, KerrNonlinearity]
+
+
+class NonlinearSpec(ABC, Tidy3dBaseModel):
+    """Abstract specification for adding nonlinearities to a medium.
+
+    Note
+    ----
+    The nonlinear constitutive relation is solved iteratively; it may not converge
+    for strong nonlinearities. Increasing `num_iters` can help with convergence.
+
+    Example
+    -------
+    >>> nonlinear_susceptibility = NonlinearSusceptibility(chi3=1)
+    >>> nonlinear_spec = NonlinearSpec(models=[nonlinear_susceptibility])
+    >>> medium = Medium(permittivity=2, nonlinear_spec=nonlinear_spec)
+    """
+
+    models: Tuple[NonlinearModelType, ...] = pd.Field(
+        (),
+        title="Nonlinear models",
+        description="The nonlinear models present in this nonlinear spec. "
+        "Nonlinear models of different types are additive. "
+        "Multiple nonlinear models of the same type are not allowed.",
+    )
+
+    num_iters: pd.PositiveInt = pd.Field(
+        1,
+        title="Number of iterations",
+        description="Number of iterations for solving nonlinear constitutive relation.",
+    )
+
+    @pd.validator("models", always=True)
+    def _no_duplicate_models(cls, val):
+        """Ensure each type of model appears at most once."""
+        if val is None:
+            return val
+        models = [model.__class__ for model in val]
+        models_unique = set(models)
+        if len(models) != len(models_unique):
+            raise ValidationError(
+                "Multiple 'NonlinearModels' of the same type "
+                "were found in a single 'NonlinearSpec'. Please ensure that "
+                "each type of 'NonlinearModel' appears at most once in a single 'NonlinearSpec'."
+            )
+        return val
+
+    @pd.validator("num_iters", always=True)
+    def _validate_num_iters(cls, val, values):
+        """Check that num_iters is not too large."""
+        if val > NONLINEAR_MAX_NUMITERS:
+            raise ValidationError(
+                "'NonlinearSpec.numiters' must be less than "
+                f"{NONLINEAR_MAX_NUMITERS}, currently {val}."
+            )
+        # also check consistency with old numiters, for backwards compatibility
+        if values.get("models") is not None:
+            for model in values["models"]:
+                if isinstance(model, NonlinearSusceptibility) and model.numiters is not None:
+                    log.warning(
+                        "The API for the number of nonlinear iterations has changed. "
+                        "The old usage 'NonlinearSusceptibility.numiters' is deprecated "
+                        "and will be removed in a future release. The new usage is "
+                        "'NonlinearSpec.num_iters'."
+                    )
+                    return model.numiters
```

Comment:
so this validator will seemingly override the global `NonlinearSpec.num_iters` with the `NonlinearSusceptibility` model's `numiters`? I'm a bit confused about this.  Could you explain what this does?

Reply:
this was confusing. I removed it. `AbstractMedium._nonlinear_num_iters` now handles all the backwards compatibility stuff.